File: StudyGamePrograming06/MeshActors.cpp
#include "MeshActors.h"
#include "MeshComponent.h"
#include "MoveComponent.h"
#include "Game.h"
#include "Renderer.h"
#include "Actor.h"

MeshActors::MeshActors(Game* game)
{
	Actor* a;
	std::string meshfile;
	// サイコロ
	meshfile = "Assets/Dice.gpmesh";
	a = new Dice(game, meshfile);		// BasicMesh
	a->SetPosition(Vector3(500.0f, -200.0f, 0.0f));
	a->SetScale(100.0f);
	MeshComponent* mc = new MeshComponent(a);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/Dice.gpmesh"));
	MoveComponent* m = new MoveComponent(a);

	a = new Dice2(game);		// LambertMesh

	a = new Dice3(game);		// PhongMesh

	// 球
	a = new Sphere(this);		// BasicMesh
	a = new Sphere2(this);		// LambertMesh
	a = new Sphere3(this);		// PhongMesh

	// レーシングカー
	a = new RacingCar(this);		// BasicMesh
	a = new RacingCar2(this);		// LambertMesh
	a = new RacingCar3(this);		// PhongMesh

	// 壁と床
	a = new Plane(this);		// PhongMesh

}

Dice::Dice(Game* game) :Actor(game)
{
	SetPosition(Vector3(500.0f, -200.0f, 0.0f));
	SetScale(100.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/Dice.gpmesh"));
	mMoveComp = new MoveComponent(this);
}



Dice::Dice(Game* game, std::string meshfile) : Actor(game)
{
}

void Dice::UpdateActor(float deltaTime)
{
	Actor::UpdateActor(deltaTime);

	Vector3 axis = Vector3::Normalize(Vector3(1.0f, -1.0f, 1.0f));
	float rotSpeedMax = 30.0f * Math::Pi;
	float rotSpeed = rotSpeedMax * deltaTime;
	mMoveComp->SetRotSpeed(rotSpeed * axis);
}

Dice2::Dice2(Game* game) :Actor(game)
{
	SetPosition(Vector3(500.0f, 0.0f, 0.0f));
	SetScale(100.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/Dice2.gpmesh"));
	mMoveComp = new MoveComponent(this);
}

void Dice2::UpdateActor(float deltaTime)
{
	Actor::UpdateActor(deltaTime);

	Vector3 axis = Vector3::Normalize(Vector3(1.0f, -1.0f, 1.0f));
	float rotSpeedMax = 30.0f * Math::Pi;
	float rotSpeed = rotSpeedMax * deltaTime;
	mMoveComp->SetRotSpeed(rotSpeed * axis);
}

Dice3::Dice3(Game* game) :Actor(game)
{
	SetPosition(Vector3(500.0f, 200.0f, 0.0f));
	SetScale(100.0f);
	MeshComponent* mc = new MeshComponent(this);
	mc->SetMesh(game->GetRenderer()->GetMesh("Assets/Dice3.gpmesh"));
	mMoveComp = new MoveComponent(this);
}

void Dice3::UpdateActor(float deltaTime)
{
	Actor::UpdateActor(deltaTime);

	Vector3 axis = Vector3::Normalize(Vector3(1.0f, -1.0f, 1.0f));
	float rotSpeedMax = 30.0f * Math::Pi;
	float rotSpeed = rotSpeedMax * deltaTime;
	mMoveComp->SetRotSpeed(rotSpeed * axis);
}

File: StudyGamePrograming06_VB/MeshActors.vb
Imports System.Drawing
Imports OpenTK.Mathematics
Imports OpenTK.Windowing.GraphicsLibraryFramework
Public Class MeshActors
    Inherits Actor
    Sub New(ByRef game As Game)
        MyBase.New(game)
        Dim a As Actor
        Dim meshfile As String
		' サイコロ
		meshfile = "Assets/Dice.gpmesh"     ' BasicMesh	
		a = New Dice(game, meshfile)
		a.SetPosition(New Vector3(700.0, 500.0, 0.0))

		meshfile = "Assets/Dice2.gpmesh"        ' LambertMesh
		a = New Dice(game, meshfile)
		a.SetPosition(New Vector3(700.0, 0.0F, 0.0))

		meshfile = "Assets/Dice3.gpmesh"        ' PhongMesh
		a = New Dice(game, meshfile)
		a.SetPosition(New Vector3(700.0, -500.0F, 0.0))

		' 球
		meshfile = "Assets/Sphere.gpmesh"       ' BasicMesh	
		a = New Sphere(game, meshfile)
		a.SetPosition(New Vector3(-500.0F, 700.0F, 0.0))

		meshfile = "Assets/Sphere2.gpmesh"      ' LambertMesh
		a = New Sphere(game, meshfile)
		a.SetPosition(New Vector3(0.0, 700.0, 0.0))

		meshfile = "Assets/Sphere3.gpmesh"      ' PhongMesh
		a = New Sphere(game, meshfile)
		a.SetPosition(New Vector3(500.0, 700.0, 0.0))

		' レーシングカー
		meshfile = "Assets/RacingCar.gpmesh"        ' BasicMesh	
		a = New RacingCar(game, meshfile)
		a.SetPosition(New Vector3(500.0, -700.0, -100.0))

		meshfile = "Assets/RacingCar2.gpmesh"       ' LambertMesh
		a = New RacingCar(game, meshfile)
		a.SetPosition(New Vector3(0.0, -700.0, -100.0))

		meshfile = "Assets/RacingCar3.gpmesh"       ' PhongMesh
		a = New RacingCar(game, meshfile)
		a.SetPosition(New Vector3(-500.0, -700.0, -100.0))

		' 壁と床
		meshfile = "Assets/Plane3.gpmesh"       ' PhongMesh
		Dim start As Single = -1250.0
		Dim size As Single = 250.0
		' 10個ずつ縦横に並べる
		For i As Integer = 0 To 9
			For j As Integer = 0 To 9
				a = New Plane(game, meshfile)
				a.SetPosition(New Vector3(start + i * size, start + j * size, -100.0))
			Next
		Next
		' 左右壁を作成
		Dim q = Quaternion.FromAxisAngle(Vector3.UnitX, -0.5 * Math.PI)
		For i As Integer = 0 To 9
			a = New Plane(game, meshfile)
			a.SetPosition(New Vector3(start + i * size, start - size, 0.0))
			a.SetRotation(q)
		Next
		q = Quaternion.FromAxisAngle(Vector3.UnitX, 0.5 * Math.PI)
		For i As Integer = 0 To 9
			a = New Plane(game, meshfile)
			a.SetPosition(New Vector3(start + i * size, -start + size, 0.0))
			a.SetRotation(q)
		Next

		' 前後壁を作成
		q = Quaternion.Multiply(Quaternion.FromAxisAngle(Vector3.UnitZ, 0.5 * Math.PI), q)
		For i As Integer = 0 To 9
			a = New Plane(game, meshfile)
			a.SetPosition(New Vector3(start - size, start + i * size, 0.0))
			a.SetRotation(q)
		Next
		q = Quaternion.Multiply(Quaternion.FromAxisAngle(Vector3.UnitZ, Math.PI), q)
		For i As Integer = 0 To 9
			a = New Plane(game, meshfile)
			a.SetPosition(New Vector3(-start + size, start + i * size, 0.0))
			a.SetRotation(q)
		Next

	End Sub
End Class

Public Class Dice
	Inherits Actor
	'public
	Sub New(ByRef game As Game, ByRef meshfile As String)
		MyBase.New(game)
		SetScale(100.0)
		Dim mc As MeshComponent = New MeshComponent(Me)
		mc.SetMesh(game.GetRenderer().GetMesh(meshfile))
		mMoveComp = New MoveComponent(Me, 20)
	End Sub

	Public Overrides Sub UpdateActor(deltaTime As Single)
		MyBase.UpdateActor(deltaTime)
		Dim axis As Vector3 = New Vector3(1.0, 1.0, 1.0)
		axis.Normalize()
		Dim rotSpeedMax As Single = -30.0 * Math.PI
		Dim rotSpeed As Single = rotSpeedMax * deltaTime
		mMoveComp.SetRotSpeed(rotSpeed * axis)
	End Sub

	'private
	Dim mMoveComp As MoveComponent

End Class

Public Class Plane
	Inherits Actor
	'public
	Sub New(ByRef game As Game, ByRef meshfile As String)
		MyBase.New(game)
		SetScale(10.0)
		Dim mc As MeshComponent = New MeshComponent(Me)
		mc.SetMesh(game.GetRenderer.GetMesh(meshfile))
	End Sub
End Class

Public Class RacingCar
	Inherits Actor
	'public
	Sub New(ByRef game As Game, ByRef meshfile As String)
		MyBase.New(game)
		SetScale(1.0)
		Dim mc = New MeshComponent(Me)
		mc.SetMesh(game.GetRenderer.GetMesh(meshfile))
		mMoveComp = New MoveComponent(Me, 20)
	End Sub
	Public Overrides Sub UpdateActor(deltaTime As Single)
		MyBase.UpdateActor(deltaTime)
		Dim w As Single = Math.PI / 4
		'Dim r As Single = Math.Sqrt(GetPosition.X * GetPosition.X + GetPosition.Y * GetPosition.Y)
		'Dim x As Single = GetPosition.X * Math.Cos(w * deltaTime) + GetPosition.Y * Math.Sin(w * deltaTime)
		'Dim y As Single = GetPosition.Y * Math.Cos(w * deltaTime) - GetPosition.X * Math.Sin(w * deltaTime)
		'SetPosition(New Vector3(x, y, -100.0))
		Dim q As Quaternion = GetRotation()
		Dim inc As Quaternion = Quaternion.FromAxisAngle(Vector3.UnitZ, -w * deltaTime)
		q = Quaternion.Multiply(inc, q)
		SetRotation(q)
	End Sub
	'private
	Dim mMoveComp As MoveComponent
End Class

Public Class Sphere
	Inherits Actor
	'public
	Sub New(ByRef game As Game, ByRef meshfile As String)
		MyBase.New(game)
		Dim q As Quaternion = New Quaternion(Vector3.UnitX, -23.5 / 180.0 * Math.PI)
		SetRotation(q)
		SetScale(5.0)
		Dim mc As MeshComponent = New MeshComponent(Me)
		mc.SetMesh(game.GetRenderer().GetMesh(meshfile))
		mMoveComp = New MoveComponent(Me, 20)
	End Sub

	Public Overrides Sub UpdateActor(deltaTime As Single)
		MyBase.UpdateActor(deltaTime)
		Dim mAxis As Vector3 = New Vector3(0.0, Math.Cos(66.5 / 180.0 * Math.PI), Math.Sin(66.5 / 180.0 * Math.PI))
		Dim rotSpeedMax As Single = 5.0 * Math.PI
		Dim rotSpeed As Single = rotSpeedMax * deltaTime
		mMoveComp.SetRotSpeed(rotSpeed * mAxis)
	End Sub

	'private
	Dim mMoveComp As MoveComponent
End Class